ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt

```diff
@@ -4,32 +4,35 @@
 
 package io.ktor.utils.io
 
-import io.ktor.utils.io.internal.*
 import io.ktor.utils.io.locks.*
 import kotlinx.atomicfu.*
 import kotlinx.coroutines.*
 import kotlinx.io.*
-import kotlin.concurrent.*
+import kotlin.concurrent.Volatile
 import kotlin.coroutines.*
+import kotlin.jvm.*
 
 @InternalAPI
-public val CHANNEL_MAX_SIZE: Int = 4096
+public const val CHANNEL_MAX_SIZE: Int = 1024 * 1024
```

Comment:
So wrapping the continuation with the local type increases the memory allocations when waiting for I/O a little bit, although we can avoid a lot of suspension just by increasing our buffer size here - this also seems to increase performance a bit too.  From what I can tell we don't read our files fully into memory anywhere, so moving the flush to the end would be problematic with large files.

Since the underlying buffer segments are recycled, I think increasing our general buffer size will generally (and counterintuitively) reduce memory overhead.

Reply:
lgtm